src/LanguageServer/Impl/CodeActions/MissingImportCodeActionProvider.cs

```diff
@@ -0,0 +1,635 @@
+                yield return fullName;
+            }
+
+            List<string> OrderImportNames(IEnumerable<string> fullNames) {
+                return fullNames.OrderBy(n => n, ImportNameComparer.Instance).ToList();
+            }
+
+            bool FilterImportedMembers(KeyValuePair<string, ImportInfo> kv) => kv.Value.Symbol.MemberType != PythonMemberType.Module && kv.Value.MemberImported;
+            bool FilterImportedModuleMembers(KeyValuePair<string, ImportInfo> kv) => kv.Value.Symbol.MemberType != PythonMemberType.Module && kv.Value.ModuleImported;
+            bool FilterImportedModules(KeyValuePair<string, ImportInfo> kv) => kv.Value.Symbol.MemberType == PythonMemberType.Module && kv.Value.MemberImported;
+
+            bool FilterSourceDeclarations(IGrouping<LocationInfo, KeyValuePair<string, ImportInfo>> group) {
+                var count = 0;
+                foreach (var entry in group) {
+                    if (count++ > 0) {
+                        return false;
+                    }
+
+                    var value = entry.Value;
+                    if (value.ModuleImported || value.ModuleImported) {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+        }
+
+        private static async Task EnsureCandidateModulesLoadedAsync(IDocumentAnalysis analysis,
+                                                                    string name,
+                                                                    CancellationToken cancellationToken) {
+            var indexManager = analysis.ExpressionEvaluator.Services.GetService<IIndexManager>();
+            if (indexManager == null) {
+                // indexing is not supported
+                return;
+            }
+
+            var symbolsIncludingName = await indexManager.WorkspaceSymbolsAsync(name, maxLength: int.MaxValue, includeLibraries: true, cancellationToken);
+
+            // we only consider exact matches rather than partial matches
+            var symbolsWithName = symbolsIncludingName.Where(s => s.Name == name && s.Kind != Indexing.SymbolKind.Variable);
+
+            var analyzer = analysis.ExpressionEvaluator.Services.GetService<IPythonAnalyzer>();
+            var pathResolver = analysis.Document.Interpreter.ModuleResolution.CurrentPathResolver;
+
+            var modules = ImmutableArray<IPythonModule>.Empty;
+            foreach (var moduleName in symbolsWithName.Select(s => s.DocumentPath).Distinct().Select(p => pathResolver.GetModuleNameByPath(p))) {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var module = analysis.Document.Interpreter.ModuleResolution.GetOrLoadModule(moduleName);
+                if (module.Analysis is EmptyAnalysis) {
+                    // once module is analyzed, this analysis result will be kept in memory
+                    // even if graph for current document (modified) changed.
```

Comment:
As above, loading modules not explicitly use is heavy, both time and memory-wise. IMO, module source can be parsed and indexed via workspace indexer, but analysis of something user did not import is a no-no.

Reply:
so, how heavy will it be? so, we are going back to only offering one that is already loaded by users?

Reply:
Very heavy. Basically nothing should be loading modules for analysis unless it is an explicit user action. Rename/find all references does it in a limited fashion, typically it effectively scopes to loaded libraries and user code. Loading and analyzing libraries like `plotly` may take minutes (just this one), consume hundreds mb of memory. Attempt to load something like `conda` environment will definitely kill machine with 8GB RAM. And burn battery charge on analysis. In short - analysis is expensive action and should be avoided at all costs.

If we are to offer items from all universe then approach should be as it is for indexing workspace - parsing and lightweight identification of possible items. It is pretty easy to find top level classes, functions and variables and this probably covers 95%.

Reply:
will make a change and ping you guys again